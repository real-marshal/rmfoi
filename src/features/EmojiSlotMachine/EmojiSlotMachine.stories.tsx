import type { Meta, Story } from '@storybook/react'
import { EmojiSlotMachine, EmojiSlotMachineProps } from './EmojiSlotMachine'

export default {
  title: 'Features/EmojiSlotMachine',
  component: EmojiSlotMachine,
} as Meta

const Template: Story<EmojiSlotMachineProps> = (args) => <EmojiSlotMachine {...args} />

export const Basic = Template.bind({})
Basic.args = {}
