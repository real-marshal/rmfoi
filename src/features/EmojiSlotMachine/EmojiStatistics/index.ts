export { EmojiStatistics } from './EmojiStatistics'
export type { EmojiStatisticsProps } from './EmojiStatistics'
