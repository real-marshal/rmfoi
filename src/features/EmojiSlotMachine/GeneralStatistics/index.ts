export { GeneralStatistics } from './GeneralStatistics'
export type { GeneralStatisticsProps } from './GeneralStatistics'
