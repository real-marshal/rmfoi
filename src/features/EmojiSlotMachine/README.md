# Some afterthoughts

## About the state

I started the main logic (not animation) as just one class to encapsulate the data and methods keeping it separated from the view. But when the time came to implement the view with React it turned out I needed lots of glue in-between, effectively killing all benefits of reactive React and instead doing all that updating logic manually and duplicating the state in React.

The easy way was to just plug in MobX to make the class auto-observable and use it directly without any problems. But I've never really used MobX and can't imagine debugging some complex state there. I could've used MobX only for this mini-project but it feels weird to have both redux and MobX in the codebase so instead I decided to rewrite the logic to redux store.

Now that I look at the store.ts, tbh it feels pretty crazy how much boilerplate redux still requires, even with RTK. So I'm not really sure if this was the right decision but well at least I remembered how to work with redux lol. Of course, I could've also just implemented everything in React, but I hate mixing logic with view. Ideally, I'd like to have it all in a class the way I started, but still, it's better be dependent on redux than react.

## Regarding animation

I spent at least 1-2 days to come up with this final algorithm and make it work consistently and still I'm not sure if this is the best way of doing this. Especially the jump, it's pretty hacky, requires prepended elements and limits a bit the possible stop indexes (which made me introduce maxIndexToWin). But it works perfectly and has good performance so that's fine I guess. Besides, I don't really have any ideas of how else I could do it better. Maybe with canvas? And Pixi to get that sweet WebGL performance?
