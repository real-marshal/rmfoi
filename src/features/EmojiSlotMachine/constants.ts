export const emojisMap = {
  '🐈': 'meow meow bitch',
  '🤦‍♀️': 'facepalm',
  '💀': 'dead',
  '🤡': 'oficially a clown',
  '🚀': 'to the moon!',
  '🤓': 'too nerdy',
  '🤯': 'mind fking blown',
  '🤑': '$$$$$$$$',
  '💩': 'shiiiit',
  '😈': 'so bad',
  '👊': 'nice one bro',
  '💪': 'gigachad',
  '💅': 'yas queen',
  '🦍': 'apes together strong',
  '💦': 'aaahhhhh',
  '🍑': 'nice booty',
  '🌶': 'damn so hot',
  '🗿': 'bruh',
  '🍆': 'wow so long',
  '📈': 'stonks go up',
  '💎': 'diamond hands',
  '🙈': 'nothing to see here',
  '🍻': 'cheers',
  '🏆': 'actually won the game',
  '🎉': 'congrats',
  '👑': 'king',
} as const

export type Emoji = keyof typeof emojisMap

export const emojis = Object.keys(emojisMap) as readonly Emoji[]

export const MIN_CHANCE = 1
export const MAX_CHANCE = 100

export enum EmojiSlotMachineMode {
  EASY = 'easy',
  STANDARD = 'standard',
  HARD = 'hard',
}

export const modeData = {
  [EmojiSlotMachineMode.EASY]: {
    chance: 40,
    chanceIncrement: 5,
  },
  [EmojiSlotMachineMode.STANDARD]: {
    chance: 20,
    chanceIncrement: 5,
  },
  [EmojiSlotMachineMode.HARD]: {
    chance: 10,
    chanceIncrement: 2.5,
  },
} as const

export const loseMessages = ['loser', 'meh', 'rekt', 'lost again?', 'fail'] as const
