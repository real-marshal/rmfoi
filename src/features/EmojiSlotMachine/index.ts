export { EmojiSlotMachine } from './EmojiSlotMachine'
export type { EmojiSlotMachineProps } from './EmojiSlotMachine'
