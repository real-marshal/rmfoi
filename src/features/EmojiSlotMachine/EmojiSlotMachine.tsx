import { MouseEventHandler, useEffect, useRef, useState } from 'react'
import {
  actions,
  evaluateNewPull,
  selectEmojiArrays,
  useCurrentDispatch,
  useCurrentSelector,
} from './store'
import { useSlotMachineAnimation } from './useSlotMachineAnimation'
import { Emoji, emojis, emojisMap, loseMessages } from './constants'
import { pickRandom } from '@/utils'
import { CenterRowOverlay, Column, Columns, EmojiElement, PullButton, Wrapper } from './styles'

interface EmojiSlotMachineProps {
  elementsInColumn: number
}

interface WonPullState {
  status: 'won'
  wonEmoji: Emoji
}

interface OtherPullState {
  status: 'pending' | 'lost'
}

type PullState = WonPullState | OtherPullState

const EmojiSlotMachine = ({ elementsInColumn = 5 }: EmojiSlotMachineProps) => {
  const ref = useRef<HTMLDivElement>(null)

  const originalEmojiArrays = useCurrentSelector(selectEmojiArrays)

  const [pullState, setPullState] = useState<PullState>({ status: 'pending' })

  const dispatch = useCurrentDispatch()

  useEffect(() => {
    dispatch(
      actions.initialize({
        maxIndexToWin: emojis.length - elementsInColumn - 1,
      })
    )
  }, [dispatch, elementsInColumn])

  const {
    startAnimation,
    isBusy,
    prependedSlotsArrays: emojiArrays,
  } = useSlotMachineAnimation({
    ref,
    slotsArrays: originalEmojiArrays,
    elementsInColumn,
    slotElementQuery: '.esm__emoji',
    columnElementQuery: '.esm__column',
  })

  const pullClickHandler: MouseEventHandler = async () => {
    setPullState({ status: 'pending' })

    const { wonEmojiIndexes, wonEmoji } = dispatch(evaluateNewPull()) ?? {}

    await startAnimation(wonEmojiIndexes)

    setPullState(wonEmoji ? { status: 'won', wonEmoji } : { status: 'lost' })
  }

  return (
    <Wrapper>
      <Columns ref={ref}>
        <CenterRowOverlay pullState={pullState}>
          <span>
            {pullState.status === 'won' && emojisMap[pullState.wonEmoji]}
            {pullState.status === 'lost' && pickRandom(loseMessages)}
          </span>
        </CenterRowOverlay>
        {emojiArrays.map((shuffledEmojis, ind) => (
          <Column key={ind} className='esm__column' elementsInColumn={elementsInColumn}>
            {shuffledEmojis.map((emoji, index) => (
              <EmojiElement key={index} className='esm__emoji'>
                {emoji}
              </EmojiElement>
            ))}
          </Column>
        ))}
      </Columns>
      <PullButton onClick={pullClickHandler} disabled={isBusy}>
        Pull
      </PullButton>
    </Wrapper>
  )
}

export { EmojiSlotMachine }
export type { EmojiSlotMachineProps, PullState }
