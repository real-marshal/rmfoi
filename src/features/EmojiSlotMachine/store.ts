import store, { MakeCurrentDispatch, MakeCurrentState, MakeCurrentThunk } from '@/app/store'
import { createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'
import { Emoji, emojis, EmojiSlotMachineMode, modeData } from './constants'
import { getShuffledEmojiArrays, getWonEmojiIndexes, EmojiSwapData, tryToWin } from './utils'

const NAME = 'emojiSlotMachine' as const

interface EmojiSlotMachineState {
  mode: EmojiSlotMachineMode
  columns: number
  emojiArrays: Emoji[][]
  chance: number
  chanceIncrement: number
  minIndexToWin: number
  maxIndexToWin: number
  stats: {
    pulls: number
    wonEmojis: Partial<Record<Emoji, number>>
  }
}

// This slice will be injected asyncronously which is why we need separate types here
interface SliceState {
  [NAME]: EmojiSlotMachineState
}
export type CurrentRootState = MakeCurrentState<SliceState>
export type CurrentDispatch = MakeCurrentDispatch<SliceState>
// eslint-disable-next-line @typescript-eslint/no-invalid-void-type
export type CurrentThunk<ReturnType = void> = MakeCurrentThunk<SliceState, ReturnType>

export const useCurrentDispatch: () => CurrentDispatch = useDispatch
export const useCurrentSelector: TypedUseSelectorHook<CurrentRootState> = useSelector

const initialState: EmojiSlotMachineState = {
  mode: EmojiSlotMachineMode.STANDARD,
  columns: 3,
  emojiArrays: getShuffledEmojiArrays(3),
  minIndexToWin: 0,
  maxIndexToWin: emojis.length - 1,
  stats: {
    pulls: 0,
    wonEmojis: {},
  },
  ...modeData[EmojiSlotMachineMode.STANDARD],
}

/*

REDUCERS

*/

const { actions, reducer } = createSlice({
  name: NAME,
  initialState,
  reducers: {
    initialize: (
      state,
      action: PayloadAction<
        Partial<Pick<EmojiSlotMachineState, 'columns' | 'mode' | 'minIndexToWin' | 'maxIndexToWin'>>
      >
    ) => {
      Object.assign(state, action.payload)
    },
    setColumns: (state, action: PayloadAction<number>) => {
      state.columns = action.payload
    },
    setEmojiArrays: (state, action: PayloadAction<EmojiSlotMachineState['emojiArrays']>) => {
      state.emojiArrays = action.payload
    },
    win: (state, action: PayloadAction<Emoji>) => {
      state.stats.pulls++

      Object.assign(state, modeData[state.mode])

      const wonEmoji = state.stats.wonEmojis[action.payload]
      state.stats.wonEmojis[action.payload] = wonEmoji ? wonEmoji + 1 : 1
    },
    lose: (state) => {
      state.stats.pulls++
      state.chance += state.chanceIncrement
    },
    swapEmojis: (state, action: PayloadAction<EmojiSwapData[]>) => {
      const { emojiArrays } = state

      action.payload.forEach(
        ({ arrayIndex, emojiInd1, emojiInd2 }) =>
          ([emojiArrays[arrayIndex]![emojiInd1]!, emojiArrays[arrayIndex]![emojiInd2]!] = [
            emojiArrays[arrayIndex]![emojiInd2]!,
            emojiArrays[arrayIndex]![emojiInd1]!,
          ])
      )
    },
    updateMode: (state, action: PayloadAction<EmojiSlotMachineMode>) => {
      state.mode = action.payload
      Object.assign(state, modeData[state.mode])
    },
  },
})

export { actions }

store.injectReducer(NAME, reducer)

/*

THUNKS

*/

// shuffleEmojiArrays has impure RNG logic so it shouldn't be in the reducer
export const shuffleEmojiArrays = (): CurrentThunk => (dispatch, getState) => {
  const { columns } = getState()[NAME]
  const shuffledEmojis = getShuffledEmojiArrays(columns)

  dispatch(actions.setEmojiArrays(shuffledEmojis))
}

export const updateColumns =
  (columns: number): CurrentThunk =>
  (dispatch) => {
    dispatch(actions.setColumns(columns))
    dispatch(shuffleEmojiArrays())
  }

export const evaluateNewPull =
  (): CurrentThunk<{ wonEmojiIndexes: number[]; wonEmoji: Emoji } | undefined> =>
  (dispatch, getState) => {
    dispatch(shuffleEmojiArrays())

    const { chance, minIndexToWin, maxIndexToWin, emojiArrays } = getState()[NAME]
    const wonEmoji = tryToWin({ chance, minIndexToWin, maxIndexToWin })

    if (!wonEmoji) {
      dispatch(actions.lose())
      return
    }

    dispatch(actions.win(wonEmoji))

    const { wonEmojiIndexes, emojisToSwap } = getWonEmojiIndexes({
      emojiArrays,
      wonEmoji,
      minIndexToWin,
      maxIndexToWin,
    })
    dispatch(actions.swapEmojis(emojisToSwap))

    return { wonEmojiIndexes, wonEmoji }
  }

/*

SELECTORS

*/

export const selectEmojiArrays = (state: CurrentRootState) => state[NAME].emojiArrays

export const selectMode = (state: CurrentRootState) => state[NAME].mode

export const selectChance = (state: CurrentRootState) => state[NAME].chance

export const selectChanceIncrement = (state: CurrentRootState) => state[NAME].chanceIncrement

export const selectStats = (state: CurrentRootState) => state[NAME].stats

export const selectGeneralStats = createSelector(
  selectStats,
  selectChance,
  selectChanceIncrement,
  ({ pulls, wonEmojis }, chance, chanceIncrement) => {
    const wins = Object.values(wonEmojis).reduce((result, wins) => result + wins, 0)

    return {
      pulls,
      wins,
      loses: pulls - wins,
      WR: pulls ? wins / pulls : 0,
      chance,
      chanceIncrement,
    }
  }
)

export const selectdEmojiStats = createSelector(selectStats, ({ wonEmojis }) => {
  const wonEmojisNumber = Object.keys(wonEmojis).length

  return {
    wonEmojisNumber,
    leftEmojis: emojis.length - wonEmojisNumber,
    wonEmojis,
  }
})
