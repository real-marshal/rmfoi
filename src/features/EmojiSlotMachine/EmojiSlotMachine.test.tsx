import store from '@/app/store'
import { composeStories } from '@storybook/react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import * as stories from './EmojiSlotMachine.stories'

// eslint-disable-next-line @typescript-eslint/no-unsafe-return
jest.mock('./store', () => ({
  ...jest.requireActual('./store'),
  selectEmojiArrays: () => [
    ['ðŸš€', 'ðŸ’Ž', 'ðŸ‘‘'],
    ['ðŸš€', 'ðŸ’Ž', 'ðŸ‘‘'],
    ['ðŸš€', 'ðŸ’Ž', 'ðŸ‘‘'],
  ],
}))

describe('EmojiSlotMachine component', () => {
  const testStories = Object.values(composeStories(stories)).map(
    (Story) => [Story.storyName, Story] as [string, typeof Story]
  )

  it.each(testStories)('renders %s story', (_, Story) => {
    const view = render(
      <Provider store={store}>
        <Story />
      </Provider>
    )
    expect(view.baseElement).toMatchSnapshot()
  })
})
