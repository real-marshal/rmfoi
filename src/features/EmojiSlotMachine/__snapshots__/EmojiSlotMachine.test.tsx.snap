// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmojiSlotMachine component renders Basic story 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  gap: 10px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  gap: 10px;
}

.emotion-4 {
  position: absolute;
  width: 110%;
  height: 60px;
  top: calc(50% - 30px);
  left: -5%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 2rem;
  -webkit-text-stroke: 1px var(--dt-theme-primary);
  text-transform: uppercase;
  text-align: center;
  font-weight: bold;
  color: #ffffff;
}

.emotion-6 {
  padding: 0 10px;
  border: 2px solid var(--dt-theme-primary);
  border-radius: 10px;
  height: 200px;
  overflow: hidden;
}

.emotion-8 {
  font-size: 2rem;
  height: 40px;
}

.emotion-48 {
  text-transform: uppercase;
  font-weight: bold;
  background: #df451b;
  color: white;
  border: none;
  border-radius: 10px;
  padding: 20px;
  cursor: pointer;
  height: 60px;
  font-size: 1.2rem;
}

.emotion-48:hover {
  background: #df2222;
}

<body>
  <div>
    <div
      class="emotion-0 emotion-1"
    >
      <div
        class="emotion-2 emotion-3"
      >
        <div
          class="emotion-4 emotion-5"
        >
          <span />
        </div>
        <div
          class="esm__column emotion-6 emotion-7"
        >
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸš€
          </div>
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸ’Ž
          </div>
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸ‘‘
          </div>
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸš€
          </div>
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸ’Ž
          </div>
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸ‘‘
          </div>
        </div>
        <div
          class="esm__column emotion-6 emotion-7"
        >
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸš€
          </div>
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸ’Ž
          </div>
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸ‘‘
          </div>
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸš€
          </div>
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸ’Ž
          </div>
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸ‘‘
          </div>
        </div>
        <div
          class="esm__column emotion-6 emotion-7"
        >
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸš€
          </div>
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸ’Ž
          </div>
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸ‘‘
          </div>
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸš€
          </div>
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸ’Ž
          </div>
          <div
            class="esm__emoji emotion-8 emotion-9"
          >
            ðŸ‘‘
          </div>
        </div>
      </div>
      <button
        class="emotion-48 emotion-49"
      >
        Pull
      </button>
    </div>
  </div>
</body>
`;
